cmake/modules/llext.cmake

```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2024 Schneider Electric
+# SPDX-License-Identifier: Apache-2.0
+
+include_guard(GLOBAL)
+
+# Add a macro that add a llext extension.
+# macro is based on zephyr_library template
+#
+# output an elf file usable by llext
+#
+# Code can be linked by 2 differents ways:
+#     1. using partial link (-r flag in gcc and -fno-pic) (default)
+#     2. using shared link (-shared flag in gcc and -fpic)
+#
+# The code will be compiled with mostly the same C compiler flags used
+# in the Zephyr build, but with some important modifications.
+# Toolchain flag are automatically inherited and filtered.
+#
+# NOPIC:
+#     -mlong-calls flag is always added
+#     libc is detected from zephyr configuration and libc function code is embedded in extension
+# PIC:
+#    code is compiled and linked with -fpic flag
+#    libc is detected from zephyr configuration and libc function are called. They are external symbol for extension.
+#
+# Example usage:
+#     zephyr_llext(hello_world)
+#     zephyr_llext_sources(hello_world.c hello_world2.c)
+#     zephyr_llext_include_directories(.)
+# will compile the source file hello_world.c and hello_world2.c to
+# ${PROJECT_BINARY_DIR}/hello_world.llext
+#
+
+
+macro(zephyr_llext name)
+    set(LLEXT_IS_PIC no)
+    set(ZEPHYR_CURRENT_LLEXT ${name})
+    set(ZEPHYR_LLEXT_LINKER_SCRIPT ${ZEPHYR_BASE}/include/zephyr/linker/llext.ld)
+
+    # ARGN is not a variable: assign its value to a variable
+    set(ExtraMacroArgs ${ARGN})
+    # Get the length of the list
+    list(LENGTH ExtraMacroArgs NumExtraMacroArgs)
+    # Execute the following block only if the length is > 0
+    if(NumExtraMacroArgs GREATER 0)
+        foreach(ExtraArg ${ExtraMacroArgs})
+            if(${ExtraArg} STREQUAL PIC)
+                set(LLEXT_IS_PIC yes)
+            endif()
+        endforeach()
+    endif()
+
+    add_executable(${ZEPHYR_CURRENT_LLEXT})
+
+    set_property(TARGET ${ZEPHYR_CURRENT_LLEXT} APPEND PROPERTY LINK_DEPENDS ${ZEPHYR_LLEXT_LINKER_SCRIPT})
+
+    target_compile_definitions(${ZEPHYR_CURRENT_LLEXT}
+        PRIVATE $<TARGET_PROPERTY:zephyr_interface,INTERFACE_COMPILE_DEFINITIONS>
+        PRIVATE ZEPHYR_LLEXT
+    )
+
+    target_compile_options(${ZEPHYR_CURRENT_LLEXT}
+        PRIVATE $<TARGET_PROPERTY:zephyr_interface,INTERFACE_COMPILE_OPTIONS>
+    )
+
+    if(LLEXT_IS_PIC)
+        target_compile_options(${ZEPHYR_CURRENT_LLEXT} PRIVATE -fpic -fpie)
```

Comment:
I know this code in general is GNU centric, but we actually have compiler flags defined for such cases to avoid custom toolchain flags / handling throughout the codebase.

This makes it easier to add better support for more toolchains in Zephyr.

For example:
https://github.com/zephyrproject-rtos/zephyr/blob/9c05618c2b6b8e10ab523411067e32846cd8cd8e/cmake/compiler/gcc/compiler_flags.cmake#L232-L234